### PR TITLE
Dynamically load case details when clicking markers

### PR DESCRIPTION
**Pull request by MartinRiese**
## Product Description

Instead of getting the content of the map marker pop ups from a case list column get it from the details.
The change actually supports both ways for the transition. If the address popup is not defined in a case list column the code will attempt to fetch it from the details. That means the app needs to be updated to define the address popup field in the case details instead of the case list. Any subsequent click on the marker will not fetch the details again.

## Technical Summary

https://dimagi-dev.atlassian.net/browse/USH-3195

## Feature Flag

* USH: Allow use of a map in the case list in Web Apps …

### Safety story

Tested locally:

1. Previously configured app with marker pop upstill works
2. Removing popup config from case list does not show any popup
3. Adding new popup config in case details triggers details request and shows popup
4. After changing popup content using clickable icons and reopening the popup will show new content
5. Adding multiple taps in case details still shows popup the same way.

### Automated test coverage

Not sure how to test that automatically since most of it involves a third party library.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js [hunk 1]
```diff
@@ -252,10 +252,12 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
             obj.header = headers[i];
             obj.style = styles[i];
             obj.id = i;
-            if (obj.style.displayFormat === 'Markdown') {
+            if (obj.style.displayFormat === constants.FORMAT_MARKDOWN) {
                 obj.html = markdown.render(details[i]);
             }
-            detailModel.push(obj);
+            if (obj.style.displayFormat !== constants.FORMAT_ADDRESS_POPUP) {
```
Comment (MartinRiese) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:258]: No super sure about this. For the case list we set widthHint to -1 in FP to hide fields. But for details they are -1. Either way to me it seems more a UI responsibility to hide it.

**Comment by MartinRiese** [comment]
Any suggestions for automated test are welcome.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 2]
```diff
@@ -869,10 +869,11 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
                                         const fetchingDetails = FormplayerFrontend.getChannel().request("entity:get:details", urlObject, false, false, false, true);
                                         $.when(fetchingDetails).done(function (detailResponse) {
                                             const attributes = Array.from(detailResponse)[0].attributes;
-                                            const popupIndex = _.findIndex(attributes.styles, (style) => style.displayFormat === constants.FORMAT_ADDRESS_POPUP);
-                                            if (popupIndex >= 0) {
-                                                const popupText = markdown.render(attributes.details[popupIndex]);
-                                                const p = L.popup().setContent(popupText);
+                                            const popupIndexOnClick =
```
Comment (Robert-Costello) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:872]: By shadowing do you mean the same variable names potentially getting used in other places?

**Comment by Robert-Costello** [comment]
> Any suggestions for automated test are welcome.

It would be a relatively big effort to create some mock mapbox responses similar to the way we mock formplayer responses. It might be something to consider as a separate ticket or series of tickets in the future. https://github.com/dimagi/commcare-hq/blob/3a66161257ea8c7e8e969506a73d9481d8d50abe/corehq/apps/cloudcare/static/cloudcare/js/formplayer/spec/fake_formplayer.js#L2

**Review by Robert-Costello — APPROVED** [review]
Left one question, but this looks good to me. Will this go on staging for testing?

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js (hunk 2 above)
Comment (MartinRiese) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:872]: Yes same variable name as used in outer scope.

**Comment by Jtang-1** [comment]
I'm guessing the continued support for address popup configuration in the case list will be just temporary? If so, is the transition plan to just ask app builders to move the configuration?

**Comment by MartinRiese** [comment]
@Jtang-1 So far it has been possible to use the Address Popup format in both case list and detail. Although, only in case list it caused the popup so far. I will investigate adding validation to not use it on the list.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js [hunk 3]
```diff
@@ -863,6 +863,23 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
                                 }
 
                                 marker.on('click', () => {
+                                    if (!marker.getPopup()) {
+                                        const caseId = model.get('id');
+                                        const urlObject = formplayerUtils.currentUrlToObject();
+                                        urlObject.addSelection(caseId);
+                                        const fetchingDetails = FormplayerFrontend.getChannel().request("entity:get:details", urlObject, false, false, false, true);
```
Comment (nospame) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:869]: Doesn't this function only take 4 parameters?
https://github.com/dimagi/commcare-hq/blob/79f02b5e0ec6be0839a230f547cf6be7334162c3/corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js#L227

**Review by nospame — COMMENTED** [review]
re: @Jtang-1 's point, based on the added validator it seems like the plan _is_ to move to only the new case detail format for address popup, right? Is there a followup planned or will the `address_popup_in_long` validator stay in place indefinitely?

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js (hunk 3 above)
Comment (MartinRiese) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:869]: Good catch. I just copied another call I found and modified it. @Robert-Costello was there a reason you used 5 arguments [here](https://github.com/dimagi/commcare-hq/blob/23943f827d095de9a25a02703e3ca668fad8348a/corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js#L333)?

Comment (Robert-Costello) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:869]: It was for a flag that indicated that the request was coming from a clickable icon. It should have been removed when we went a different route with case list filtering. If you want you could remove the extra argument here or I can make a separate PR.

**Comment by Jtang-1** [comment]
It sounds like the plan is to
With this PR, 
- continue to support existing builds where address pop up is defined in case list. 
- don't allow new builds where address popup is defined in case list (Does address popup need to be removed from "format" options on case list UI?)

Followup PR 
- removes the backend code that supports address popups defined in case list
- How do we ensure all builds have been converted to using case detail for address popup?

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js (hunk 3 above)
Comment (MartinRiese) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js]: I removed them

**Comment by MartinRiese** [comment]
@Jtang-1 I think it really is only used in the BHA apps and woody knows about the change. So I am certain he will track it.

File: corehq/apps/app_manager/helpers/validators.py [hunk 4]
```diff
@@ -565,10 +565,13 @@ def validate_case_list_field_actions(self):
 class ModuleDetailValidatorMixin(object):
 
     __invalid_tile_configuration_type: str = "invalid tile configuration"
-
     __invalid_clickable_icon_configuration: str = "invalid clickable icon configuration"
+    __deprecated_popup_configuration: str = "deprecated popup configuration"
```
Comment (Jtang-1) [inline corehq/apps/app_manager/helpers/validators.py:569]: Is this used anywhere?

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js [hunk 5]
```diff
@@ -179,7 +179,12 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
     };
 
     var showDetail = function (model, detailTabIndex, caseId, isMultiSelect) {
-        var detailObjects = model.models;
+        var detailObjects = model.filter(function(d){
+            const styles = d.get('styles');
+            const visibleStyle = _.find(styles, s => s.displayFormat !== constants.FORMAT_ADDRESS_POPUP);
```
Comment (Jtang-1) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:184]: could `styles` potentially be undefined?

**Review by Jtang-1 — APPROVED** [review]

**Comment by Jtang-1** [comment]
Also noting - the validation will make it such that apps currently using address-popup on case list will not be able to make a new build without moving it to case details. I'm assuming that's been communicated to delivery

File: corehq/apps/app_manager/helpers/validators.py (hunk 4 above)
Comment (MartinRiese) [reply corehq/apps/app_manager/helpers/validators.py:569]: I meant to. Good catch

File: corehq/apps/app_manager/helpers/validators.py [hunk 6]
```diff
@@ -585,6 +588,20 @@ def _validate_fields_with_format(
                             .format(format_display, fields_with_address_format_str))
             })
 
+    def _validate_address_popup_in_long(
+        self,
+        errors: list
+    ):
+        fields_with_address_format = \
+            [c.field for c in self.module.case_details.short.columns if c.format == self.__address_popup]
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/helpers/validators.py:596]: yes. changed it.

**Comment by MartinRiese** [comment]
@Jtang-1 @nospame Change of plans. I disabled the validator for now and added a migration command. Since the UI code supports address popup in both the case list and case detail, we can do the deployment in the following steps.

1. Deploy
2. Run the command
3. Wait a week. Hopefully everybody that needs to creates a new version or their app
4. Next deploy
  * Add validator back in
  * Possibly remove the code that support address popups in case lists
  * Add an actual migration that runs the command but will skip all already processed projects

**Review by Jtang-1 — APPROVED** [review]
I see there's a bunch of code commented out to be reapplied after the migration. My preference is would be to just delete the code and put it in another branch. But other than that, looks good.

**Pull request merged**